src/_gettsim/aggregation_numpy.py

```diff
@@ -127,6 +127,18 @@ def grouped_cumsum(column, group_id):
     return out
 
 
+def sum_values_by_index(column, id_col, p_id):
+    fail_if_dtype_not_numeric_or_boolean(column, agg_func="sum_values_by_index")
+    if column.dtype in ["bool"]:
+        column = column.astype(int)
+    out = numpy.zeros_like(p_id, dtype=column.dtype)
+
+    for position, id_receiver in enumerate(id_col):
+        if id_receiver >= 0:
+            out[numpy.where(p_id == id_receiver)] += column[position]
```

Comment:
Performance can be improved here: Currently, you have an outer loop (for..in) going over the $n$ elements of `id_col` and an inner loop (vectorized `==`) going over the $n$ elements of `p_id`. Thus, you currently have quadratic complexity, i.e. doubling $n$ quadruples the runtime.

You are better off first creating a `dict` mapping `p_id`s to positions (linear complexity) and then using this dictionary to look up the position of the person with `p_id == id_receiver` for each entry in `id_col` (linear complexity). This overall still has linear complexity instead of quadratic.

Reply:
Makes sense, thanks!